File: src/helpers/ptx-trace/sdk/examples/java/EtlPipeline.java
package ai.portunix.trace.examples;

import ai.portunix.trace.*;
import ai.portunix.trace.cli.CliException;
import ai.portunix.trace.models.*;

import java.util.*;
import java.util.regex.Pattern;

/**
 * PTX-TRACE Java SDK Example - ETL Pipeline
 *
 * This example demonstrates a complete ETL pipeline with tracing.
 * It shows how to:
 * - Create and manage sessions
 * - Trace extraction, transformation, and loading phases
 * - Use source information for data lineage
 * - Handle errors with proper severity
 * - Use fluent API for tracing
 */
public class EtlPipeline {

    // Sample data
    private static final String[][] SAMPLE_DATA = {
        {"1", "John Doe", "john@example.com", "+420 777 123 456"},
        {"2", "Jane Smith", "jane@test", "+1 555 234 5678"},
        {"3", "Bob Wilson", "bob@company.org", "invalid"},
        {"4", "Alice Brown", "alice@example.com", "+44 20 1234 5678"}
    };

    private static final String[] COLUMNS = {"id", "name", "email", "phone"};

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("PTX-TRACE Java SDK - ETL Pipeline Example");
        System.out.println("=".repeat(60));

        try {
            System.out.println("\n1. Running ETL Pipeline...\n");
            runEtlPipeline();

            System.out.println("\n2. Running Fluent API Example...\n");
            runFluentApiExample();

            System.out.println("\n" + "=".repeat(60));
            System.out.println("All examples completed!");
            System.out.println("=".repeat(60));

        } catch (CliException e) {
            System.err.println("Error: " + e.getMessage());
            System.err.println("Exit code: " + e.getExitCode());
            System.exit(1);
        }
    }

    /**
     * Run the complete ETL pipeline with tracing.
     */
    private static void runEtlPipeline() throws CliException {
        // Create session with PII masking enabled
        try (Session session = Trace.newSession("customer-import")
                .withPIIMasking(true)
                .withTags("etl", "daily")
                .build()) {

            // ============================================
            // PHASE 1: EXTRACT
            // ============================================

            List<Map<String, String>> records;

            try (Operation op = session.start("extract")) {
                op.source(SourceInfo.csv("customers.csv"));
                op.context("format", "csv");

                // Read sample data
                records = readSampleData();

                op.output("record_count", records.size());
                op.output("columns", String.join(",", COLUMNS));
                op.success();
            }

            System.out.println("Extracted " + records.size() + " records");

            // ============================================
            // PHASE 2: TRANSFORM
            // ============================================

            List<Map<String, String>> transformedRecords = new ArrayList<>();
            int errorCount = 0;

            for (int rowNum = 0; rowNum < records.size(); rowNum++) {
                Map<String, String> record = records.get(rowNum);

                try (Operation op = session.start("transform")) {
                    op.source(SourceInfo.csv("customers.csv", rowNum + 1));
                    op.input("id", record.get("id"));
                    op.input("name", record.get("name"));
                    op.input("email", record.get("email"));
                    op.input("phone", record.get("phone"));

                    // Validate email
                    if (!validateEmail(record.get("email"))) {
                        op.error("Invalid email format: " + record.get("email"), Severity.MEDIUM);
                        op.tag("validation_error");
                        errorCount++;
                        continue;
                    }

                    // Transform record
                    try {
                        Map<String, String> transformed = transformRecord(record);

                        // Check phone normalization
                        if (transformed.get("phone") == null) {
                            op.error("Invalid phone number format", Severity.LOW);
                            // Recovery: set to null
                        }

                        op.output("id", transformed.get("id"));
                        op.output("name", transformed.get("name"));
                        op.output("email", transformed.get("email"));
                        op.output("phone", transformed.get("phone") != null ? transformed.get("phone") : "NULL");
                        op.success();

                        transformedRecords.add(transformed);

                    } catch (Exception e) {
                        op.error(e);
                        errorCount++;
                    }
                }
            }

            System.out.println("Transformed " + transformedRecords.size() + " records, " + errorCount + " errors");

            // ============================================
            // PHASE 3: LOAD
            // ============================================

            try (Operation op = session.start("load")) {
                op.context("destination", "postgres://localhost/customers");
                op.context("table", "customers");
                op.input("record_count", transformedRecords.size());

                try {
                    int inserted = loadToDatabase(transformedRecords);

                    op.output("inserted", inserted);
                    op.output("skipped", records.size() - inserted);
                    op.success();

                } catch (Exception e) {
                    op.error(e.getMessage(), Severity.CRITICAL);
                    throw e;
                }
            }

            // ============================================
            // SUMMARY
            // ============================================

            Map<String, Object> stats = session.stats();

            System.out.println("\n" + "=".repeat(50));
            System.out.println("ETL Pipeline Summary");
            System.out.println("=".repeat(50));
            System.out.println("Session ID: " + session.getId());
            System.out.println("Stats: " + stats);
        }
    }

    /**
     * Example using the fluent API.
     */
    private static void runFluentApiExample() throws CliException {
        try (Session session = Trace.createSession("fluent-example")) {

            // Fluent API with execute
            session.trace("process_data")
                .withType("etl")
                .input("source", "api")
                .input("count", 100)
                .source(SourceInfo.database("postgres://localhost/db", "users"))
                .tag("production", "batch")
                .withRule("R001", "1.0")
                .execute(ctx -> {
                    ctx.output("processed", 100);
                    ctx.output("status", "complete");
                    ctx.tag("success");
                });

            System.out.println("Fluent API example completed");
        }
    }

    /**
     * Read sample data into list of maps.
     */
    private static List<Map<String, String>> readSampleData() {
        List<Map<String, String>> records = new ArrayList<>();

        for (String[] row : SAMPLE_DATA) {
            Map<String, String> record = new LinkedHashMap<>();
            for (int i = 0; i < COLUMNS.length; i++) {
                record.put(COLUMNS[i], row[i]);
            }
            records.add(record);
        }

        return records;
    }

    /**
     * Normalize phone number by removing spaces and validating format.
     */
    private static String normalizePhone(String phone) {
        if (phone == null || phone.isEmpty()) {
            return null;
        }

        // Remove spaces
        String normalized = phone.replaceAll("\\s+", "");

        // Basic validation - must start with + and have digits
        if (!normalized.startsWith("+")) {
            return null;
        }

        String digits = normalized.substring(1).replaceAll("-", "");
        if (!digits.matches("\\d+")) {
            return null;
        }

        return normalized;
    }

    /**
     * Validate email format.
     */
    private static boolean validateEmail(String email) {
        if (email == null || email.isEmpty()) {
            return false;
        }

        int atIndex = email.indexOf("@");
        if (atIndex < 1) {
            return false;
        }

        String domain = email.substring(atIndex + 1);
        return domain.contains(".");
    }

    /**
     * Transform a single record.
     */
    private static Map<String, String> transformRecord(Map<String, String> record) {
        Map<String, String> transformed = new LinkedHashMap<>();

        transformed.put("id", record.get("id"));

        // Normalize name
        String name = record.get("name");
        if (name != null) {
            name = name.trim();
            // Title case
            String[] words = name.split("\\s+");
            StringBuilder sb = new StringBuilder();
            for (String word : words) {
                if (sb.length() > 0) sb.append(" ");
                sb.append(Character.toUpperCase(word.charAt(0)));
                if (word.length() > 1) {
                    sb.append(word.substring(1).toLowerCase());
                }
            }
            name = sb.toString();
        }
        transformed.put("name", name);

        // Normalize email
        String email = record.get("email");
        if (email != null) {
            email = email.toLowerCase().trim();
        }
        transformed.put("email", email);

        // Normalize phone
        transformed.put("phone", normalizePhone(record.get("phone")));

        return transformed;
    }

    /**
     * Simulate loading records to database.
     */
    private static int loadToDatabase(List<Map<String, String>> records) {
        System.out.println("Loading " + records.size() + " records to database...");
        // In real implementation, this would insert into database
        return records.size();
    }
}

File: src/helpers/ptx-trace/sdk/examples/typescript/etl-pipeline.ts
/**
 * PTX-TRACE TypeScript SDK Example - ETL Pipeline
 *
 * This example demonstrates a complete ETL pipeline with tracing.
 * It shows how to:
 * - Create and manage sessions
 * - Trace extraction, transformation, and loading phases
 * - Use source information for data lineage
 * - Handle errors with proper severity
 * - Use fluent API for tracing
 */

import {
  Session,
  csvSource,
  dbSource,
  Severity,
} from '../src'; // Use 'ptx-trace' when installed from npm

// Sample data
const SAMPLE_DATA = [
  { id: '1', name: 'John Doe', email: 'john@example.com', phone: '+420 777 123 456' },
  { id: '2', name: 'Jane Smith', email: 'jane@test', phone: '+1 555 234 5678' },
  { id: '3', name: 'Bob Wilson', email: 'bob@company.org', phone: 'invalid' },
  { id: '4', name: 'Alice Brown', email: 'alice@example.com', phone: '+44 20 1234 5678' },
];

interface Record {
  id: string;
  name: string;
  email: string;
  phone: string | null;
}

/**
 * Normalize phone number by removing spaces and validating format.
 */
function normalizePhone(phone: string): string | null {
  if (!phone) {
    return null;
  }

  // Remove spaces
  const normalized = phone.replace(/\s+/g, '');

  // Basic validation - must start with + and have digits
  if (!normalized.startsWith('+')) {
    return null;
  }

  const digits = normalized.substring(1).replace(/-/g, '');
  if (!/^\d+$/.test(digits)) {
    return null;
  }

  return normalized;
}

/**
 * Validate email format.
 */
function validateEmail(email: string): boolean {
  if (!email) {
    return false;
  }

  const atIndex = email.indexOf('@');
  if (atIndex < 1) {
    return false;
  }

  const domain = email.substring(atIndex + 1);
  return domain.includes('.');
}

/**
 * Transform a single record.
 */
function transformRecord(record: typeof SAMPLE_DATA[0]): Record {
  // Normalize name (title case)
  const name = record.name
    .trim()
    .split(/\s+/)
    .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join(' ');

  return {
    id: record.id,
    name,
    email: record.email.toLowerCase().trim(),
    phone: normalizePhone(record.phone),
  };
}

/**
 * Simulate loading records to database.
 */
async function loadToDatabase(records: Record[]): Promise<number> {
  console.log(`Loading ${records.length} records to database...`);
  // Simulate async operation
  await new Promise(resolve => setTimeout(resolve, 100));
  return records.length;
}

/**
 * Run the complete ETL pipeline with tracing.
 */
async function runEtlPipeline(): Promise<void> {
  // Create session with PII masking enabled
  const session = await Session.create('customer-import', {
    piiMasking: true,
    tags: ['etl', 'daily'],
  });

  try {
    // ============================================
    // PHASE 1: EXTRACT
    // ============================================

    let records: typeof SAMPLE_DATA;

    await session.withOperation('extract', async (op) => {
      op.source(csvSource('customers.csv'));
      op.context('format', 'csv');

      // Read sample data
      records = [...SAMPLE_DATA];

      op.output('record_count', records.length);
      op.output('columns', Object.keys(records[0]).join(','));
    });

    console.log(`Extracted ${records!.length} records`);

    // ============================================
    // PHASE 2: TRANSFORM
    // ============================================

    const transformedRecords: Record[] = [];
    let errorCount = 0;

    for (let rowNum = 0; rowNum < records!.length; rowNum++) {
      const record = records![rowNum];

      await session.withOperation('transform', async (op) => {
        op.source(csvSource('customers.csv', rowNum + 1));
        op.input('id', record.id);
        op.input('name', record.name);
        op.input('email', record.email);
        op.input('phone', record.phone);

        // Validate email
        if (!validateEmail(record.email)) {
          op.error(`Invalid email format: ${record.email}`, Severity.MEDIUM);
          op.tag('validation_error');
          errorCount++;
          return;
        }

        // Transform record
        try {
          const transformed = transformRecord(record);

          // Check phone normalization
          if (transformed.phone === null) {
            op.error('Invalid phone number format', Severity.LOW);
            // Recovery: set to null
          }

          op.output('id', transformed.id);
          op.output('name', transformed.name);
          op.output('email', transformed.email);
          op.output('phone', transformed.phone ?? 'NULL');

          transformedRecords.push(transformed);

        } catch (e) {
          op.error(String(e), Severity.HIGH);
          errorCount++;
        }
      });
    }

    console.log(`Transformed ${transformedRecords.length} records, ${errorCount} errors`);

    // ============================================
    // PHASE 3: LOAD
    // ============================================

    await session.withOperation('load', async (op) => {
      op.context('destination', 'postgres://localhost/customers');
      op.context('table', 'customers');
      op.input('record_count', transformedRecords.length);

      try {
        const inserted = await loadToDatabase(transformedRecords);

        op.output('inserted', inserted);
        op.output('skipped', records!.length - inserted);

      } catch (e) {
        op.error(String(e), Severity.CRITICAL);
        throw e;
      }
    });

    // ============================================
    // SUMMARY
    // ============================================

    const stats = await session.stats();

    console.log('\n' + '='.repeat(50));
    console.log('ETL Pipeline Summary');
    console.log('='.repeat(50));
    console.log(`Session ID: ${session.id}`);
    console.log(`Total Events: ${stats.stats?.totalEvents}`);
    console.log(`By Status: ${JSON.stringify(stats.stats?.byStatus)}`);
    console.log(`By Level: ${JSON.stringify(stats.stats?.byLevel)}`);

  } finally {
    await session.close();
  }
}

/**
 * Example using the fluent API.
 */
async function runFluentApiExample(): Promise<void> {
  const session = await Session.create('fluent-example');

  try {
    // Fluent API with execute
    await session.trace('process_data')
      .withType('etl')
      .input('source', 'api')
      .input('count', 100)
      .source(dbSource('postgres://localhost/db', 'users'))
      .tag('production', 'batch')
      .withRule('R001', '1.0')
      .execute(async (ctx) => {
        ctx.output('processed', 100);
        ctx.output('status', 'complete');
        ctx.tag('success');
      });

    console.log('Fluent API example completed');

  } finally {
    await session.close();
  }
}

/**
 * Example with manual operation control.
 */
async function runManualOperationExample(): Promise<void> {
  const session = await Session.create('manual-example');

  try {
    // Build operation manually
    const op = session.trace('process_batch')
      .input('batch_id', 123)
      .tag('manual')
      .build();

    try {
      // Do some work
      op.input('items', 50);

      // Simulate processing
      await new Promise(resolve => setTimeout(resolve, 50));

      op.output('processed', 50);
      op.output('failed', 0);
      op.success();

    } catch (e) {
      op.error(String(e), Severity.HIGH);
      throw e;
    } finally {
      await op.end();
    }

    console.log('Manual operation example completed');

  } finally {
    await session.close();
  }
}

// Main execution
async function main(): Promise<void> {
  console.log('='.repeat(60));
  console.log('PTX-TRACE TypeScript SDK - ETL Pipeline Example');
  console.log('='.repeat(60));

  try {
    console.log('\n1. Running ETL Pipeline...\n');
    await runEtlPipeline();

    console.log('\n2. Running Fluent API Example...\n');
    await runFluentApiExample();

    console.log('\n3. Running Manual Operation Example...\n');
    await runManualOperationExample();

    console.log('\n' + '='.repeat(60));
    console.log('All examples completed!');
    console.log('='.repeat(60));

  } catch (e) {
    console.error('Error:', e);
    process.exit(1);
  }
}

main();
